datastore/datastore/storagebackend/postgresql/getobservations.go

```diff
@@ -422,14 +433,38 @@ func getObs(db *sql.DB, request *datastore.GetObsRequest, obs *[]*datastore.Meta
 
 	obsMdatas := make(map[int64][]*datastore.ObsMetadata) // observations per time series ID
 
+	// set oldest allowable obs time when in 'latest' mode
+	oldestTime := time.Time{}
+	if !tspec.IntervalMode {
+		loTime, hiTime := common.GetValidTimeRange()
+		if tspec.LatestMaxage < 0 { // i.e. filtering on maxage is disabled
+			oldestTime = loTime
+		} else {
+			oldestTime = hiTime.Add(-tspec.LatestMaxage)
+		}
+	}
+
 	// scan rows
 	for rows.Next() {
 		obsMdata, tsID, err := scanObsRow(rows)
 		if err != nil {
 			return fmt.Errorf("scanObsRow() failed: %v", err)
 		}
 
-		obsMdatas[tsID] = append(obsMdatas[tsID], obsMdata)
+		includeObs := true // by default include obs in this time series (in particular in
+		// 'interval' mode, since filtering for that has been done already)
+		if !tspec.IntervalMode { // 'latest' mode
+			switch {
+			case len(obsMdatas[tsID]) >= tspec.LatestLimit:
+				includeObs = false // reject, since not room for this obs
+			case obsMdata.GetObstimeInstant().AsTime().Before(oldestTime):
+				includeObs = false // reject, since obs too old
+			}
```

Comment:
You're quite right about latest_maxage - it can be done in the initial SELECT, not as a post processing filter. But now we remove this feature anyway (see above comment).

I also agree with the suggestion of using SELECT DISTINCT ON. This seems to be the best implementation (in terms of both performance and code complexity) if you're only interested in the single latest observation anyway.

Change coming up!